src/cbor/serializer.rs

```diff
@@ -0,0 +1,132 @@
+use std::collections::BTreeMap;
+use cbor::cbor::{CBORType};
+
+// pub enum CborType<'a> {
+//     UInt(u64),
+//     NInt(i64),
+//     BStr(&'a [u8]),
+//     TStr(&'a String),
+//     Arr(&'a [CborType<'a>]),
+//     Map(&'a BTreeMap<i64, CborType<'a>>), // TODO: find out what key value range we really have to support
+// }
+
+/// Given a vector of bytes to append to, a tag to use, and an unsigned value to encode, uses the
+/// CBOR unsigned integer encoding to represent the given value.
+fn common_encode_unsigned(output: &mut Vec<u8>, tag: u8, value: u64) {
+    assert!(tag < 8);
+    let shifted_tag = tag << 5;
+    match value {
+        0 ... 23 => {
+            output.push(shifted_tag | (value as u8));
+        },
+        24 ... 255 => {
+            output.push(shifted_tag | 24);
+            output.push(value as u8);
+        },
+        256 ... 65535 => {
+            output.push(shifted_tag | 25);
+            output.push((value >> 8) as u8);
+            output.push((value & 255) as u8);
+        },
+        65536 ... 4294967295 => {
+            output.push(shifted_tag | 26);
+            output.push((value >> 24) as u8);
+            output.push(((value >> 16) & 255) as u8);
+            output.push(((value >> 8) & 255) as u8);
+            output.push((value & 255) as u8);
+        },
+        _ => {
+            output.push(shifted_tag | 27);
+            output.push((value >> 56) as u8);
+            output.push(((value >> 48) & 255) as u8);
+            output.push(((value >> 40) & 255) as u8);
+            output.push(((value >> 32) & 255) as u8);
+            output.push(((value >> 24) & 255) as u8);
+            output.push(((value >> 16) & 255) as u8);
+            output.push(((value >> 8) & 255) as u8);
+            output.push((value & 255) as u8);
+        }
+    };
+}
+
+/// The major type is 0. For values 0 through 23, the 5 bits of additional information is just the
+/// value of the unsigned number. For values representable in one byte, the additional information
+/// has the value 24. If two bytes are necessary, the value is 25. If four bytes are necessary, the
+/// value is 26. If 8 bytes are necessary, the value is 27. The following bytes are the value of the
+/// unsigned number in as many bytes were indicated in network byte order (big endian).
+fn encode_unsigned(output: &mut Vec<u8>, unsigned: u64) {
+    common_encode_unsigned(output, 0, unsigned);
+}
+
+/// The major type is 1. The encoding is the same as for positive (i.e. unsigned) integers, except
+/// the value encoded is -1 minus the value of the negative number.
+fn encode_negative(output: &mut Vec<u8>, negative: i64) {
+    assert!(negative < 0);
+    let value_to_encode: u64 = (-1 - negative) as u64;
+    common_encode_unsigned(output, 1, value_to_encode);
+}
+
+/// The major type is 2. The length of the data is encoded as with positive integers, followed by
+/// the actual data.
+fn encode_bstr(output: &mut Vec<u8>, bstr: &[u8]) {
```

Comment:
Right now encoder and decoder use slightly different terminology for the CBOR types (e.g. bstr vs. bytes) - I imagine it would be easier to maintain if we were consistent. This can be a follow-up clean-up patch, though.

Reply:
Right, I only changed the types, not the function names. Let's do it in a follow-up.